src/controller.rs

```diff
@@ -578,32 +599,130 @@ impl AppState {
-        )
-        .with_priority_fee(priority_fee, ctx.cu_limit)
-        .place_orders(orders)
-        .build();
+        let orders_iter = req.orders.into_iter();
+        match req.place_order_type {
+            PlaceOrderType::Tx => {
+                let orders = orders_iter
+                    .map(|o| {
+                        let base_decimals =
+                            get_market_decimals(self.client.program_data(), o.market);
+                        o.to_order_params(base_decimals)
+                    })
+                    .collect();
+                let tx = TransactionBuilder::new(
+                    self.client.program_data(),
+                    sub_account,
+                    Cow::Owned(account_data),
+                    self.wallet.is_delegated(),
+                )
+                .with_priority_fee(priority_fee, ctx.cu_limit)
+                .place_orders(orders)
+                .build();
+
+                let tx_res = self.send_tx(tx, "place_orders", ctx.ttl).await;
+                match tx_res {
+                    Ok(tx_res) => Ok(PlaceOrderResponse::Tx(tx_res)),
+                    Err(e) => Err(e),
+                }
+            }
+            PlaceOrderType::SignedMsg => {
+                let orders_len = orders_iter.len();
+                let mut signed_messages = Vec::with_capacity(orders_len);
+                let mut hashes: Vec<String> = Vec::with_capacity(orders_len);
+                let sub_account_id = ctx.sub_account_id.unwrap_or(self.default_subaccount_id);
+                let current_slot = self.slot_subscriber.current_slot();
+                let orders_with_hex: Vec<(OrderParams, Vec<u8>)> = orders_iter
+                    .map(|order| {
+                        let base_decimals =
+                            get_market_decimals(self.client.program_data(), order.market);
+                        let order_for_signing_hex = order.clone();
+                        let order_params = order.to_order_params(base_decimals);
+                        (
+                            order_params,
+                            order_for_signing_hex.to_signed_order_hex(
+                                order_params,
+                                current_slot,
+                                sub_account_id,
+                            ),
+                        )
+                    })
+                    .collect();
+
+                for (order, message) in orders_with_hex {
+                    let signature = self.wallet.sign_message(message.as_slice())?;
+                    let market_type: &'static str = match order.market_type {
+                        MarketType::Spot => "spot",
+                        MarketType::Perp => "perp",
+                    };
+                    let incoming_msg = IncomingSignedMessage {
+                        taker_authority: self.authority().to_string(),
+                        signature: general_purpose::STANDARD.encode(signature),
```

Comment:
i think this'll base64 it
```suggestion
                        signature: signature.to_string(),
```